cirq-core/cirq/qis/clifford_tableau.py

```diff
@@ -137,30 +138,91 @@ class CliffordTableau(StabilizerState):
     an eigenoperator of the state vector with eigenvalue one: P|psi> = |psi>.
     """
 
-    def __init__(self, num_qubits, initial_state: int = 0):
+    def __init__(
+        self,
+        num_qubits,
+        initial_state: int = 0,
+        rs: Optional[np.ndarray] = None,
+        xs: Optional[np.ndarray] = None,
+        zs: Optional[np.ndarray] = None,
+    ):
         """Initializes CliffordTableau
         Args:
             num_qubits: The number of qubits in the system.
             initial_state: The computational basis representation of the
                 state as a big endian int.
         """
         self.n = num_qubits
-
-        # The last row (`2n+1`-th row) is the scratch row used in _measurement
+        self.initial_state = initial_state
+        # _reconstruct_* adds the last row (`2n+1`-th row) to the input arrays,
+        # which is the scratch row used in _measurement
         # computation process only. It should not be exposed to external usage.
-        self._rs = np.zeros(2 * self.n + 1, dtype=bool)
-
-        for (i, val) in enumerate(
-            big_endian_int_to_digits(initial_state, digit_count=num_qubits, base=2)
-        ):
-            self._rs[self.n + i] = bool(val)
+        self._rs = self._reconstruct_rs(rs)
+        self._xs = self._reconstruct_xs(xs)
+        self._zs = self._reconstruct_zs(zs)
+
+    def _reconstruct_rs(self, rs: Optional[np.ndarray]) -> np.ndarray:
+        if rs is None:
+            new_rs = np.zeros(2 * self.n + 1, dtype=bool)
+            for (i, val) in enumerate(
+                big_endian_int_to_digits(self.initial_state, digit_count=self.n, base=2)
+            ):
+                new_rs[self.n + i] = bool(val)
+        else:
+            shape = rs.shape
+            if len(shape) == 1 and shape[0] == 2 * self.n and rs.dtype == np.dtype(bool):
+                new_rs = np.append(rs, np.zeros(1, dtype=bool))
+            else:
+                raise ValueError(
+                    f"The value you passed for rs is not the correct shape and/or type. "
+                    f"Please confirm that it's 1 row, of even length and of type bool."
```

Comment:
It isn't enough for the length to be even. It actually needs to be twice the number of qubits. Suggestion:

```python
f"Please ensure that it's a single row with two bool entries per qubit."
```

(The message in your other exception, raised on line 197, already reflects this.)

Reply:
Done.